pkg/firewall/firewall.go

```diff
@@ -0,0 +1,151 @@
+package firewall
+
+import (
+	"crypto/ecdsa"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/ipfs/go-log"
+
+	coreChain "github.com/keep-network/keep-core/pkg/chain"
+	coreFirewall "github.com/keep-network/keep-core/pkg/firewall"
+	coreNet "github.com/keep-network/keep-core/pkg/net"
+	coreKey "github.com/keep-network/keep-core/pkg/net/key"
+
+	eth "github.com/keep-network/keep-ecdsa/pkg/chain"
+)
+
+var logger = log.Logger("keep-ecdsa-firewall")
+
+// KeepCacheLifetime is the time the cache maintains the list of active keep
+// members. We use the cache to minimize calls to Ethereum client.
+const KeepCacheLifetime = 24 * time.Hour
+
+var errNoMinStakeNoActiveKeep = fmt.Errorf("remote peer has no minimum " +
+	"stake and is not a member in any of active keeps")
+
+// NewStakeOrActiveKeepPolicy if a firewall policy checking if the remote peer
+// has a minimum stake and if it has no minimum stake if it is a member in at
+// least one active keep.
+func NewStakeOrActiveKeepPolicy(
+	chain eth.Handle,
+	stakeMonitor coreChain.StakeMonitor,
+) coreNet.Firewall {
+	return &stakeOrActiveKeepPolicy{
+		chain:                  chain,
+		minimumStakePolicy:     coreFirewall.MinimumStakePolicy(stakeMonitor),
+		activeKeepMembersCache: newTimeCache(KeepCacheLifetime),
+	}
+}
+
+type stakeOrActiveKeepPolicy struct {
+	chain                  eth.Handle
+	minimumStakePolicy     coreNet.Firewall
+	activeKeepMembersCache *timeCache
+}
+
+func (soakp *stakeOrActiveKeepPolicy) Validate(
+	remotePeerPublicKey *ecdsa.PublicKey,
+) error {
+	// Validate minimum stake policy. If the remote peer has the minimum stake,
+	// we are fine and we should let to connect.
+	if err := soakp.minimumStakePolicy.Validate(remotePeerPublicKey); err == nil {
+		return nil
+	}
+
+	// In case the remote peer has no minimum stake, we need to check if it is
+	// a member in at least one active keep. If so, we let to connect.
+	// Otherwise, we do not let to connect.
+	return soakp.validateActiveKeepMembership(remotePeerPublicKey)
+}
+
+func (soakp *stakeOrActiveKeepPolicy) validateActiveKeepMembership(
+	remotePeerPublicKey *ecdsa.PublicKey,
+) error {
+	remotePeerNetworkPublicKey := coreKey.NetworkPublic(*remotePeerPublicKey)
+	remotePeerAddress := coreKey.NetworkPubKeyToEthAddress(&remotePeerNetworkPublicKey)
```

Comment:
Do we need this `remote` prefix? 

```suggestion
	peerPublicKey *ecdsa.PublicKey,
) error {
	peerNetworkPublicKey := coreKey.NetworkPublic(*peerPublicKey)
	peerAddress := coreKey.NetworkPubKeyToEthAddress(&peerNetworkPublicKey)
```

Reply:
It's a convention we use in the network layer to indicate it's not the current's peer property.